### DISCUSSION TITLE
Test "Should ask gameMap to drop shape when touching the ground upon step, and clear full rows."

### DISCUSSION THREAD

**Discussion by PrazTobie** [web-classroom/web-classroom.github.io#5] [Q&A : Pratique / Labos] [answered]
Je ne comprends pas le résultat attendu pour ce test. 
J'ai effectué le test graphiquement et mon implémentation ajoutes les 2 pièces dans la grille lorsqu'elle touchent le sol. 
Je trouve le comportement logique (la pièce du joueur 1 posée en premier puis la pièce du joueur 2 est posée immédiatement après).

Le test apparement ne s'attend que à ce que la pièce du joueur 1 soit posée. Si c'est le cas que se passe-t-il avec la pièce du joueur 2 étant donnée qu'elle n'est pas superposée avec celle du joueur 1 et qu'elle n'est pas supposée être posée?

**Comment by theookage** [comment]
j'ai la même question :)

**Comment by JarodStreck** [comment]
Même problème et même question. je comprends pas pourquoi le test s'attend à n'avoir qu'une pièce de posée.

**Comment by OlivierLmr** [comment, marked as answer]
Très bonne question, c'est effectivement un des tests tricky du projet.

La situation que le test vérifie est la suivante. On part d'un board avec deux shapes des joueurs 1 et 2 (pas encore posées, donc "tombantes") comme suit :
```
 222
  2 
 111
  1 
-----
```

On fait ensuite avancer le jeu d'un step. La shape 1 ne peut pas descendre, mais la shape 2 le peut. La shape 2 descend donc, et la shape 1 est posée. La shape 2 est maintenant bloquée et doit être remplacée par une nouvelle shape en haut du board.

La raison pour laquelle il doit se passer ça, c'est que l'inverse (et probablement ce qui se passe dans votre solution actuelle si vous ne passez pas ce test) serait non-déterministe. J'imagine que vous posez la shape 1 dès que vous voyez qu'elle devrait l'être, puis seulement ensuite considérez la shape 2, qui veut donc être posée elle aussi. Le problème est que, si on intervertit les shapes 1 et 2 (ou si votre code, qui utilise certainement une map, venait à itérer sur les shapes dans l'ordre inverse, en commençant par la shape 2), alors la shape du dessus, testée en premier, pourrait descendre, puis la shape 1, testée ensuite, serait posée. Dans ce cas là, la shape du dessus serait superposée à celle du dessous puis remplacée. En fonction de laquelle est considérée en premier (celle du haut ou celle du bas), on aurait donc deux outcomes différents. Ce test vérifie qu'on n'a pas ce problème et s'assurant qu'une shape qui doit être posée n'impacte pas les tests des shapes testées après elle.

Je vais rajouter une précision à l'énoncé car il est vrai qu'on pourrait arguer qu'une autre solution à ce problème serait de d'abord poser toutes les shapes à poser, et seulement ensuite faire descendre toutes celles qui le peuvent. Merci de la remarque.
